src/order/filter/filter_product.lean

```diff
@@ -0,0 +1,293 @@
+def bigly_equal : setoid (α → β) := 
+⟨ λ a b, {n | a n = b n} ∈ φ,
+  λ a, by simp only [eq_self_iff_true, (set.univ_def).symm, univ_sets], 
+  λ a b ab, by simpa only [eq_comm], 
+  λ a b c ab bc, sets_of_superset φ (inter_sets φ ab bc) (λ n r, eq.trans r.1 r.2)⟩
+
+/-- Ultraproduct, but on a general filter -/
+def filterprod := quotient (bigly_equal β φ)
+local notation `β*` := filterprod β φ
+
+namespace filter_product
+
+variables {α β φ} include φ
+
+def of_seq : (α → β) → β* := @quotient.mk' (α → β) (bigly_equal β φ)
+
+/-- Equivalence class containing the constant sequence of a term in β -/
+def of (b : β): β* := of_seq (function.const α b)
+
+/-- Lift function to filter product -/
+def lift (f : β → β) : β* → β* :=
+  λ x, quotient.lift_on' x (λ a, (of_seq $ λ n, f (a n) : β*)) $
+  λ a b h, quotient.sound' $ show _ ∈ _, by filter_upwards [h] λ i hi, congr_arg _ hi
+
+/-- Lift binary operation to filter product -/
+def lift₂ (f : β → β → β) : β* → β* → β* :=
+  λ x y, quotient.lift_on₂' x y (λ a b, (of_seq $ λ n, f (a n) (b n) : β*)) $
+  λ a₁ a₂ b₁ b₂ h1 h2, quotient.sound' $ show _ ∈ _,
+  by filter_upwards [h1, h2] λ i hi1 hi2, congr (congr_arg _ hi1) hi2
+
+/-- Lift properties to filter product -/
+def lift_rel (R : β → Prop) : β* → Prop :=
+  λ x, quotient.lift_on' x (λ a, {i : α | R (a i)} ∈ φ) $ λ a b h, propext 
+  ⟨ λ ha, by filter_upwards [h, ha] λ i hi hia, by simpa [hi.symm],
+    λ hb, by filter_upwards [h, hb] λ i hi hib, by simpa [hi.symm.symm] ⟩
+
+/-- Lift binary relations to filter product -/
+def lift_rel₂ (R : β → β → Prop) : β* → β* → Prop :=
+  λ x y, quotient.lift_on₂' x y (λ a b, {i : α | R (a i) (b i)} ∈ φ) $
+  λ a₁ a₂ b₁ b₂ h₁ h₂, propext 
+  ⟨ λ ha, by filter_upwards [h₁, h₂, ha] λ i hi1 hi2 hia, by simpa [hi1.symm, hi2.symm],
+    λ hb, by filter_upwards [h₁, h₂, hb] λ i hi1 hi2 hib, by simpa [hi1.symm.symm, hi2.symm.symm] ⟩
+
+theorem of_inj (NT : φ ≠ ⊥): function.injective (@of _ β φ) :=
+begin
+  intros r s rs, by_contra N,
+  rw [of, of, of_seq, quotient.eq', bigly_equal] at rs, 
+  simp only [N, set.set_of_false, empty_in_sets_eq_bot] at rs,
+  exact NT rs
+end
+
+theorem of_seq_fun (f g : α → β) (h : β → β) (H : {n : α | f n = h (g n) } ∈ φ) : 
+of_seq f = (lift h) (@of_seq _ _ φ g) := quotient.sound' H
+
+theorem of_seq_fun₂ (f g₁ g₂ : α → β) (h : β → β → β) (H : {n : α | f n = h (g₁ n) (g₂ n) } ∈ φ) : 
+of_seq f = (lift₂ h) (@of_seq _ _ φ g₁) (@of_seq _ _ φ g₂) := quotient.sound' H
+
+instance coe_filterprod : has_coe β β* := ⟨ of ⟩
+
+instance [has_add β] : has_add β* := { add := lift₂ has_add.add }
+
+instance [has_zero β] : has_zero β* := { zero := of 0 }
+
+instance [has_neg β] : has_neg β* :=
+{ neg := lift has_neg.neg }
+
+instance [add_semigroup β] : add_semigroup β* :=
+{ add_assoc := λ x y z, quotient.induction_on₃' x y z $ λ a b c, quotient.sound' $
+    show {n | _+_+_ = _+(_+_)} ∈ _, by simp only [add_assoc, eq_self_iff_true]; exact φ.2, 
```

Comment:
```suggestion
    show {n | _ + _ + _ = _ + (_ + _)} ∈ _, by simp only [add_assoc, eq_self_iff_true]; exact φ.univ_sets, 
```
Please use spaces around operators. Also especially for structures use `.univ_sets` instead of `.2`. Sometimes it makes sense to use `.1` or `.2` especially for products, conjunction proofs, etc. But for structures like `filter` it is too unpredictable what is at the second place.